sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala

```diff
@@ -734,15 +736,290 @@ case class DescribeTableCommand(
       case Some(spec) =>
-        spec.toLinkedHashMap.foreach(s => append(buffer, s._1, s._2, ""))
+        toLinkedHashMap(spec.toJsonLinkedHashMap)
+          .foreach(s => append(buffer, s._1, s._2, ""))
       case _ =>
     }
-    table.storage.toLinkedHashMap.foreach(s => append(buffer, s._1, s._2, ""))
+    toLinkedHashMap(table.storage.toJsonLinkedHashMap)
+      .foreach(s => append(buffer, s._1, s._2, ""))
+  }
+}
+
+/**
+ * Command that looks like
+ * {{{
+ *   DESCRIBE [EXTENDED|FORMATTED] table_name partitionSpec? [AS JSON];
+ * }}}
+ */
+case class DescribeTableJsonCommand(
+   table: TableIdentifier,
+   partitionSpec: TablePartitionSpec,
+   isExtended: Boolean)
+  extends LeafRunnableCommand {
+  override val output = DescribeCommandSchema.describeJsonTableAttributes()
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val result = new ArrayBuffer[Row]
+    val catalog = sparkSession.sessionState.catalog
+
+    if (catalog.isTempView(table)) {
+      if (partitionSpec.nonEmpty) {
+        throw QueryCompilationErrors.descPartitionNotAllowedOnTempView(table.identifier)
+      }
+      val schema = catalog.getTempViewOrPermanentTableMetadata(table).schema
+      describeColsJson(schema, result, header = false)
+    } else {
+      val metadata = catalog.getTableRawMetadata(table)
+      val schema = if (metadata.schema.isEmpty) {
+        // In older versions of Spark,
+        // the table schema can be empty and should be inferred at runtime.
+        sparkSession.table(metadata.identifier).schema
+      } else {
+        metadata.schema
+      }
+
+      addKeyValueToJson(result, "table_name", JString(metadata.identifier.table))
+
+      table.catalog.foreach(catalog => addKeyValueToJson(result, "catalog_name", JString(catalog)))
+      table.database.foreach { db =>
+        addKeyValueToJson(result, "namespace", JArray(List(JString(db))))
+        addKeyValueToJson(result, "schema_name", JString(db))
+      }
+
+      describeColsJson(schema, result, header = false)
+
+      describeClusteringInfoJson(metadata, result)
+
+      if (partitionSpec.nonEmpty) {
+        // Outputs the partition-specific info for the DDL command:
+        // "DESCRIBE [EXTENDED|FORMATTED] table_name PARTITION (partitionVal*)"
+        describePartitionInfoJson(sparkSession, catalog, metadata, result)
+      } else {
+        describeFormattedTableInfoJson(metadata, result)
+      }
+    }
+
+    result.toSeq
+  }
+
+  /**
+   * Util to recursively form JSON string representation of data type, used for DESCRIBE AS JSON.
+   * Differs from `json` in DataType.scala by providing additional fields for some types.
+   */
+  def jsonType(dataType: DataType): String = {
+    dataType match {
+      case arrayType: ArrayType =>
+        val elementTypeJson = jsonType(arrayType.elementType)
+        s"""{"type": "array","elementType": $elementTypeJson,
+           |"containsNull": ${arrayType.containsNull}}""".stripMargin
```

Comment:
we are generating the final string here, shall we put them in one line?

Reply:
It ends up being 1 line because when we add to the JSON result we get the [compact](https://github.com/apache/spark/pull/49139/files#diff-632efff0ffceb2ff47b97f39570b4f2cff24401f132f559158acaa651b27af26R925) JSON string, I split the lines here because of the scalastyle character length limit